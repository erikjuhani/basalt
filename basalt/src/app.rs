use crate::markdown::{MarkdownView, MarkdownViewState};
use basalt_core::obsidian::{Note, Vault};
use crossterm::event::{self, Event, KeyEvent, KeyEventKind};
use ratatui::{
    buffer::Buffer,
    layout::{Constraint, Layout, Rect, Size},
    widgets::{StatefulWidget, StatefulWidgetRef},
    DefaultTerminal,
};

use std::{cell::RefCell, io::Result, marker::PhantomData};

use crate::{
    config::{Config, Key},
    help_modal::{HelpModal, HelpModalState},
    sidepanel::{SidePanel, SidePanelState},
    start::{StartScreen, StartState},
    statusbar::{StatusBar, StatusBarState},
    text_counts::{CharCount, WordCount},
    vault_selector_modal::{VaultSelectorModal, VaultSelectorModalState},
};

const VERSION: &str = env!("CARGO_PKG_VERSION");

const HELP_TEXT: &str = include_str!("./help.txt");

#[derive(Debug, Clone, Default, PartialEq)]
pub enum Mode {
    #[default]
    Select,
    Normal,
    Insert,
}

impl Mode {
    fn as_str(&self) -> &'static str {
        match self {
            Mode::Select => "Select",
            Mode::Normal => "Normal",
            Mode::Insert => "Insert",
        }
    }
}

#[derive(Debug, Default, Clone, Copy, PartialEq)]
pub enum ScrollAmount {
    #[default]
    One,
    HalfPage,
}

fn calc_scroll_amount(scroll_amount: ScrollAmount, size: Size) -> usize {
    match scroll_amount {
        ScrollAmount::One => 1,
        ScrollAmount::HalfPage => (size.height / 3).into(),
    }
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub enum Action {
    Select,
    Next,
    Prev,
    Insert,
    Resize(Size),
    ScrollUp(ScrollAmount),
    ScrollDown(ScrollAmount),
    ToggleMode,
    ToggleHelp,
    ToggleVaultSelector,
    Quit,
}

#[derive(Debug, Default, Clone, PartialEq)]
pub struct Start<'a> {
    pub start_state: StartState<'a>,
}

#[derive(Debug, Default, Clone, PartialEq)]
pub struct Main<'a> {
    pub sidepanel_state: SidePanelState<'a>,
    pub selected_note: Option<SelectedNote>,
    pub markdown_view_state: MarkdownViewState,
    pub notes: Vec<Note>,
    pub vaults: Vec<&'a Vault>,
    pub size: Size,
    pub mode: Mode,
}

impl<'a> Main<'a> {
    fn new(vault_name: &'a str, notes: Vec<Note>, size: Size, vaults: Vec<&'a Vault>) -> Self {
        Self {
            notes: notes.clone(),
            sidepanel_state: SidePanelState::new(vault_name, notes),
            vaults,
            size,
            ..Default::default()
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Screen<'a> {
    Start(Start<'a>),
    Main(Main<'a>),
}

impl Default for Screen<'_> {
    fn default() -> Self {
        Screen::Start(Start::default())
    }
}

#[derive(Debug, Default, Clone, PartialEq)]
pub struct AppState<'a> {
    pub config: Config,
    pub help_modal: Option<HelpModalState>,
    pub vault_selector_modal: Option<VaultSelectorModalState<'a>>,
    pub size: Size,
    pub is_running: bool,
    pub screen: Screen<'a>,
    _lifetime: PhantomData<&'a ()>,
}

pub struct App<'a> {
    pub state: AppState<'a>,
    terminal: RefCell<DefaultTerminal>,
}

impl<'a> StatefulWidgetRef for App<'a> {
    type State = AppState<'a>;

    fn render_ref(&self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let screen = state.screen.clone();

        match screen {
            Screen::Start(mut state) => {
                StartScreen::default().render_ref(area, buf, &mut state.start_state)
            }
            Screen::Main(mut state) => {
                let [content, statusbar] =
                    Layout::vertical([Constraint::Fill(1), Constraint::Length(1)])
                        .horizontal_margin(1)
                        .areas(area);

                let (left, right) = if state.mode == Mode::Select {
                    (Constraint::Length(35), Constraint::Fill(1))
                } else {
                    (Constraint::Length(5), Constraint::Fill(1))
                };

                let [sidepanel, note] = Layout::horizontal([left, right]).areas(content);

                SidePanel::default().render_ref(sidepanel, buf, &mut state.sidepanel_state);

                MarkdownView.render_ref(note, buf, &mut state.markdown_view_state);

                let mode = state.mode.as_str().to_uppercase();
                let (name, counts) = state
                    .selected_note
                    .clone()
                    .map(|note| {
                        let content = note.content.as_str();
                        (
                            note.name,
                            (WordCount::from(content), CharCount::from(content)),
                        )
                    })
                    .unzip();

                let (word_count, char_count) = counts.unwrap_or_default();

                let mut status_bar_state = StatusBarState::new(
                    &mode,
                    name.as_deref(),
                    word_count.into(),
                    char_count.into(),
                );

                StatusBar::default().render_ref(statusbar, buf, &mut status_bar_state);
            }
        }

        if let Some(mut vault_selector_modal_state) = state.vault_selector_modal.clone() {
            VaultSelectorModal::default().render(area, buf, &mut vault_selector_modal_state)
        }

        if let Some(mut help_modal_state) = state.help_modal.clone() {
            HelpModal.render(area, buf, &mut help_modal_state)
        }
    }
}

#[derive(Debug, Default, Clone, PartialEq)]
pub struct SelectedNote {
    name: String,
    path: String,
    content: String,
}

impl From<&Note> for SelectedNote {
    fn from(value: &Note) -> Self {
        Self {
            name: value.name.clone(),
            path: value.path.to_string_lossy().to_string(),
            content: Note::read_to_string(value).unwrap(),
        }
    }
}

fn help_text() -> String {
    let version = format!("{VERSION}~alpha");
    HELP_TEXT.replace(
        "%version-notice",
        format!("This is the read-only release of Basalt ({version})").as_str(),
    )
}

impl<'a> App<'a> {
    pub fn start(terminal: DefaultTerminal, vaults: Vec<&Vault>) -> Result<()> {
        let version = format!("{VERSION}~alpha");
        let size = terminal.size()?;

        let state = AppState {
            config: Config::build(),
            screen: Screen::Start(Start {
                start_state: StartState::new(&version, size, vaults),
            }),
            size,
            is_running: true,
            _lifetime: PhantomData,
            ..Default::default()
        };

        App {
            state: state.clone(),
            terminal: RefCell::new(terminal),
        }
        .run(state)
    }

    fn run(&mut self, state: AppState<'a>) -> Result<()> {
        self.draw(&state)?;
        let event = &event::read()?;
        match self.update(&state, self.handle_event(event)) {
            state if state.is_running => self.run(state),
            _ => Ok(()),
        }
    }

    fn update_help_modal(
        &self,
        state: AppState<'a>,
        inner: HelpModalState,
        action: Option<Action>,
    ) -> AppState<'a> {
        match action {
            Some(Action::ScrollUp(amount)) => AppState {
                help_modal: Some(inner.scroll_up(calc_scroll_amount(amount, state.size))),
                ..state
            },
            Some(Action::ScrollDown(amount)) => AppState {
                help_modal: Some(inner.scroll_down(calc_scroll_amount(amount, state.size))),
                ..state
            },
            Some(Action::Next) => AppState {
                help_modal: Some(inner.scroll_down(1)),
                ..state
            },
            Some(Action::Prev) => AppState {
                help_modal: Some(inner.scroll_up(1)),
                ..state
            },
            _ => state,
        }
    }

    fn update_vault_selector_modal(
        &self,
        state: AppState<'a>,
        inner: VaultSelectorModalState<'a>,
        action: Option<Action>,
    ) -> AppState<'a> {
        match action {
            Some(Action::ToggleVaultSelector) => AppState {
                vault_selector_modal: None,
                ..state
            },
            Some(Action::Select) => {
                // TODO: Add logic to not load the vault again if the same vault was picked in the
                // selector.
                let alphabetically =
                    |a: &Note, b: &Note| a.name.to_lowercase().cmp(&b.name.to_lowercase());

                let vault_selector_state = inner.vault_selector_state.select();

                let vault_with_notes = vault_selector_state
                    .selected()
                    .and_then(|index| inner.vault_selector_state.get_item(index))
                    .map(|vault| (vault, vault.notes_sorted_by(alphabetically)));

                if let Some((vault, notes)) = vault_with_notes {
                    AppState {
                        screen: Screen::Main(Main::new(
                            &vault.name,
                            notes,
                            state.size,
                            vault_selector_state.items(),
                        )),
                        vault_selector_modal: None,
                        ..state
                    }
                } else {
                    state
                }
            }
            Some(Action::Next) => AppState {
                vault_selector_modal: Some(VaultSelectorModalState {
                    vault_selector_state: inner.vault_selector_state.next(),
                }),
                ..state
            },
            Some(Action::Prev) => AppState {
                vault_selector_modal: Some(VaultSelectorModalState {
                    vault_selector_state: inner.vault_selector_state.previous(),
                }),
                ..state
            },
            _ => state,
        }
    }

    fn update_select_mode(
        &self,
        state: AppState<'a>,
        inner: Main<'a>,
        action: Option<Action>,
    ) -> AppState<'a> {
        match action {
            Some(Action::ToggleMode) => AppState {
                screen: Screen::Main(Main {
                    mode: Mode::Normal,
                    sidepanel_state: inner.sidepanel_state.close(),
                    ..inner
                }),
                ..state
            },
            Some(Action::ScrollUp(amount)) => AppState {
                screen: Screen::Main(Main {
                    markdown_view_state: inner
                        .markdown_view_state
                        .scroll_up(calc_scroll_amount(amount, state.size)),
                    ..inner
                }),
                ..state
            },
            Some(Action::ScrollDown(amount)) => AppState {
                screen: Screen::Main(Main {
                    markdown_view_state: inner
                        .markdown_view_state
                        .scroll_down(calc_scroll_amount(amount, state.size)),
                    ..inner
                }),
                ..state
            },
            Some(Action::Select) => {
                let sidepanel_state = inner.sidepanel_state.select();

                let selected_note = inner
                    .notes
                    .get(sidepanel_state.selected().unwrap_or_default())
                    .map(SelectedNote::from);

                AppState {
                    screen: Screen::Main(Main {
                        sidepanel_state,
                        selected_note: selected_note.clone(),
                        markdown_view_state: inner
                            .markdown_view_state
                            .set_text(selected_note.map(|note| note.content).unwrap_or_default())
                            .reset_scrollbar(),
                        ..inner
                    }),
                    ..state
                }
            }
            Some(Action::Next) => AppState {
                screen: Screen::Main(Main {
                    sidepanel_state: inner.sidepanel_state.next(),
                    ..inner
                }),
                ..state
            },
            Some(Action::Prev) => AppState {
                screen: Screen::Main(Main {
                    sidepanel_state: inner.sidepanel_state.previous(),
                    ..inner
                }),
                ..state
            },
            _ => state,
        }
    }

    fn update_normal_mode(
        &self,
        state: AppState<'a>,
        inner: Main<'a>,
        action: Option<Action>,
    ) -> AppState<'a> {
        let Some(action) = action else {
            return state;
        };

        match action {
            Action::ToggleMode => AppState {
                screen: Screen::Main(Main {
                    mode: Mode::Select,
                    sidepanel_state: inner.sidepanel_state.open(),
                    ..inner
                }),
                ..state
            },
            Action::ScrollUp(amount) => AppState {
                screen: Screen::Main(Main {
                    markdown_view_state: inner
                        .markdown_view_state
                        .scroll_up(calc_scroll_amount(amount, state.size)),
                    ..inner
                }),
                ..state
            },
            Action::ScrollDown(amount) => AppState {
                screen: Screen::Main(Main {
                    markdown_view_state: inner
                        .markdown_view_state
                        .scroll_down(calc_scroll_amount(amount, state.size)),
                    ..inner
                }),
                ..state
            },
            Action::Next => AppState {
                screen: Screen::Main(Main {
                    markdown_view_state: inner.markdown_view_state.scroll_down(1),
                    ..inner
                }),
                ..state
            },
            Action::Prev => AppState {
                screen: Screen::Main(Main {
                    markdown_view_state: inner.markdown_view_state.scroll_up(1),
                    ..inner
                }),
                ..state
            },
            _ => state,
        }
    }

    fn update_main_state(
        &self,
        state: AppState<'a>,
        inner: Main<'a>,
        action: Option<Action>,
    ) -> AppState<'a> {
        if let Some(Action::ToggleVaultSelector) = action {
            return AppState {
                vault_selector_modal: if state.vault_selector_modal.is_some() {
                    None
                } else {
                    Some(VaultSelectorModalState::new(inner.vaults.clone()))
                },
                ..state
            };
        }

        match inner.mode {
            Mode::Select => self.update_select_mode(state, inner, action),
            Mode::Normal => self.update_normal_mode(state, inner, action),
            Mode::Insert => state,
        }
    }

    fn update_start_state(
        &self,
        state: AppState<'a>,
        inner: Start<'a>,
        action: Option<Action>,
    ) -> AppState<'a> {
        match action {
            Some(Action::Select) => {
                let alphabetically =
                    |a: &Note, b: &Note| a.name.to_lowercase().cmp(&b.name.to_lowercase());

                let splash_state = inner.start_state.select();

                let vault_with_notes = splash_state
                    .selected()
                    .and_then(|index| splash_state.get_item(index))
                    .map(|vault| (vault, vault.notes_sorted_by(alphabetically)));

                if let Some((vault, notes)) = vault_with_notes {
                    AppState {
                        screen: Screen::Main(Main::new(
                            &vault.name,
                            notes,
                            state.size,
                            inner.start_state.items(),
                        )),
                        ..state
                    }
                } else {
                    state
                }
            }
            Some(Action::Next) => AppState {
                screen: Screen::Start(Start {
                    start_state: inner.start_state.next(),
                }),
                ..state
            },
            Some(Action::Prev) => AppState {
                screen: Screen::Start(Start {
                    start_state: inner.start_state.previous(),
                }),
                ..state
            },
            _ => state,
        }
    }

    fn update(&self, state: &AppState<'a>, action: Option<Action>) -> AppState<'a> {
        let state = state.clone();
        let screen = state.screen.clone();
        match action {
            Some(Action::Quit) => AppState {
                is_running: false,
                ..state
            },
            Some(Action::ToggleHelp) => AppState {
                help_modal: if state.help_modal.is_some() {
                    None
                } else {
                    Some(HelpModalState::new(&help_text()))
                },
                ..state
            },
            Some(Action::Resize(size)) => AppState { size, ..state },
            _ if state.help_modal.is_some() => {
                self.update_help_modal(state.clone(), state.help_modal.unwrap().clone(), action)
            }
            _ if state.vault_selector_modal.is_some() => self.update_vault_selector_modal(
                state.clone(),
                state.vault_selector_modal.unwrap().clone(),
                action,
            ),
            _ => match screen {
                Screen::Start(inner) => self.update_start_state(state, inner, action),
                Screen::Main(inner) => self.update_main_state(state, inner, action),
            },
        }
    }

    fn handle_event(&self, event: &Event) -> Option<Action> {
        match event {
            Event::Resize(cols, rows) => Some(Action::Resize(Size::new(*cols, *rows))),
            Event::Key(key_event) if key_event.kind == KeyEventKind::Press => {
                self.handle_press_key_event(key_event)
            }
            _ => None,
        }
    }

    fn handle_press_key_event(&self, key_event: &KeyEvent) -> Option<Action> {
        let key_binding: Key = key_event.into();

        self.state
            .config
            .keymap
            .get(&key_binding)
            .cloned()
            .or_else(|| key_binding.eq(&Key::CTRLC).then_some(Action::Quit))
    }

    fn draw(&self, state: &AppState<'a>) -> Result<()> {
        let mut terminal = self.terminal.borrow_mut();
        let mut state = state.clone();

        terminal.draw(move |frame| {
            let area = frame.area();
            let buf = frame.buffer_mut();
            self.render_ref(area, buf, &mut state);
        })?;

        Ok(())
    }
}
