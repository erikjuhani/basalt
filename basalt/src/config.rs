mod key_binding;

use core::fmt;
use std::{collections::BTreeMap, fs::read_to_string};

use etcetera::{choose_base_strategy, home_dir, BaseStrategy};
use key_binding::{Command, KeyBinding};
use serde::Deserialize;

use crate::app::Message;
pub(crate) use key_binding::Key;

#[derive(Debug, thiserror::Error)]
pub enum ConfigError {
    // Standard IO error, from [`std::io::Error`].
    #[error(transparent)]
    Io(#[from] std::io::Error),
    // Occurs when the home directory cannot be located, from [`etcetera::HomeDirError`].
    #[error(transparent)]
    HomeDir(#[from] etcetera::HomeDirError),
    /// TOML (De)serialization error, from [`toml::de::Error`].
    #[error(transparent)]
    Toml(#[from] toml::de::Error),
    #[error("Invalid keybinding: {0}")]
    InvalidKeybinding(String),
    #[error("Unknown code: {0}")]
    UnknownKeyCode(String),
    #[error("Unknown modifiers: {0}")]
    UnknownKeyModifiers(String),
    #[error("User config not found: {0}")]
    UserConfigNotFound(String),
}

#[derive(Clone, Debug, PartialEq)]
pub struct ConfigSection {
    pub key_bindings: BTreeMap<String, Message>,
}

impl ConfigSection {
    /// Takes self and another config and merges the `key_bindings` together overwriting the
    /// existing entries with the value from another config.
    pub(crate) fn merge_key_bindings(&mut self, config: Self) {
        config.key_bindings.into_iter().for_each(|(key, message)| {
            self.key_bindings.insert(key, message);
        });
    }

    pub fn key_to_message(&self, key: Key) -> Option<Message> {
        self.key_bindings.get(&key.to_string()).cloned()
    }
}

impl fmt::Display for ConfigSection {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self.key_bindings
            .iter()
            .try_for_each(|(key, message)| -> fmt::Result {
                writeln!(f, "{}: {:?}", key, message)
            })?;

        Ok(())
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct Config {
    pub global: ConfigSection,
    pub splash: ConfigSection,
    pub explorer: ConfigSection,
    pub help_modal: ConfigSection,
    pub note_viewer: ConfigSection,
    pub vault_selector_modal: ConfigSection,
}

impl Default for Config {
    fn default() -> Self {
        Self::from(TomlConfig::default())
    }
}

impl From<TomlConfig> for Config {
    fn from(value: TomlConfig) -> Self {
        Self {
            global: value.global.into(),
            splash: value.splash.into(),
            explorer: value.explorer.into(),
            help_modal: value.help_modal.into(),
            note_viewer: value.note_viewer.into(),
            vault_selector_modal: value.vault_selector_modal.into(),
        }
    }
}

impl From<TomlConfigSection> for ConfigSection {
    fn from(TomlConfigSection { key_bindings }: TomlConfigSection) -> Self {
        Self {
            key_bindings: key_bindings
                .into_iter()
                .map(|KeyBinding { key, command }| (key.to_string(), command.into()))
                .collect(),
        }
    }
}

impl Config {
    /// Takes self and another config and merges the `key_bindings` together overwriting the
    /// existing entries with the value from another config.
    pub(crate) fn merge(&mut self, config: Self) -> Self {
        self.global.merge_key_bindings(config.global);
        self.explorer.merge_key_bindings(config.explorer);
        self.splash.merge_key_bindings(config.splash);
        self.note_viewer.merge_key_bindings(config.note_viewer);
        self.help_modal.merge_key_bindings(config.help_modal);
        self.vault_selector_modal
            .merge_key_bindings(config.vault_selector_modal);
        self.clone()
    }
}

impl fmt::Display for Config {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        writeln!(f, "[global]\n{}", self.global)?;
        writeln!(f, "[splash]\n{}", self.splash)?;
        writeln!(f, "[explorer]\n{}", self.explorer)?;
        writeln!(f, "[note_viewer]\n{}", self.note_viewer)?;
        writeln!(f, "[help_modal]\n{}", self.help_modal)?;
        writeln!(f, "[vault_selector_modal]\n{}", self.vault_selector_modal)?;

        Ok(())
    }
}

impl From<BTreeMap<String, Message>> for ConfigSection {
    fn from(value: BTreeMap<String, Message>) -> Self {
        Self {
            key_bindings: value,
        }
    }
}

impl<const N: usize> From<[(String, Message); N]> for ConfigSection {
    fn from(value: [(String, Message); N]) -> Self {
        BTreeMap::from(value).into()
    }
}

#[derive(Clone, Debug, PartialEq, Deserialize, Default)]
struct TomlConfigSection {
    #[serde(default)]
    key_bindings: KeyBindings,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Default)]
struct KeyBindings(Vec<KeyBinding>);

impl IntoIterator for KeyBindings {
    type Item = KeyBinding;
    type IntoIter = std::vec::IntoIter<Self::Item>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}

impl AsRef<Vec<KeyBinding>> for KeyBindings {
    fn as_ref(&self) -> &Vec<KeyBinding> {
        &self.0
    }
}

impl<const N: usize> From<[(Key, Command); N]> for KeyBindings {
    fn from(value: [(Key, Command); N]) -> Self {
        Self(value.into_iter().map(KeyBinding::from).collect())
    }
}

#[derive(Clone, Debug, PartialEq, Deserialize, Default)]
struct TomlConfig {
    #[serde(default)]
    global: TomlConfigSection,
    #[serde(default)]
    splash: TomlConfigSection,
    #[serde(default)]
    explorer: TomlConfigSection,
    #[serde(default)]
    help_modal: TomlConfigSection,
    #[serde(default)]
    note_viewer: TomlConfigSection,
    #[serde(default)]
    vault_selector_modal: TomlConfigSection,
}

/// Finds and reads the user configuration file in order of priority.
///
/// The function checks two standard locations:
///
/// 1. Directly under the user's home directory: `$HOME/.basalt.toml`
/// 2. Under the user's config directory: `$HOME/.config/basalt/config.toml`
///
/// It first attempts to find the config file in the home directory. If not found, it then checks
/// the config directory.
fn read_user_config() -> Result<Config, ConfigError> {
    let home_dir_path = home_dir().map(|home_dir| home_dir.join(".basalt.toml"));
    let config_dir_path =
        choose_base_strategy().map(|strategy| strategy.config_dir().join("basalt/config.toml"));

    let config_path = [home_dir_path, config_dir_path]
        .into_iter()
        .flatten()
        .find(|path| path.exists())
        .ok_or(ConfigError::UserConfigNotFound(
            "Could not find user config".to_string(),
        ))?;

    toml::from_str::<TomlConfig>(&read_to_string(config_path)?)
        .map(Config::from)
        .map_err(ConfigError::from)
}

const BASE_CONFIGURATION_STR: &str =
    include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/config.toml"));

/// Loads and merges configuration from multiple sources in priority order.
///
/// The configuration is built by layering sources with increasing precedence:
/// 1. Base configuration from embedded config.toml (lowest priority)
/// 2. User-specific configuration from user's config directory
/// 3. System overrides (Ctrl+C) that cannot be changed by users (highest priority)
///
/// # Configuration Precedence
/// System overrides > User config > Base config
pub fn load() -> Result<Config, ConfigError> {
    // TODO: Use compile time toml parsing instead to check the build error during compile time
    // Requires a custom proc-macro workspace crate
    let mut base_config: Config = toml::from_str::<TomlConfig>(BASE_CONFIGURATION_STR)?.into();

    // TODO: Parsing errors related to the configuration file should ideally be surfaced as warnings.
    // This is pending a solution for toast notifications and proper warning/error logging.
    if let Ok(user_config) = read_user_config() {
        base_config.merge(user_config);
    }

    let system_key_binding_overrides: ConfigSection =
        [(Key::CTRL_C.to_string(), Message::Quit)].into();

    base_config
        .global
        .merge_key_bindings(system_key_binding_overrides);

    Ok(base_config)
}

#[cfg(test)]
mod tests {
    use super::*;
    // use insta::assert_snapshot;

    #[test]
    fn test_base_config_snapshot() {
        // TODO: Does not work cross-platform as macOS has different names for the keys
        // Potentially needs two snapshots
        //
        // let config: Config = toml::from_str::<TomlConfig>(BASE_CONFIGURATION_STR)
        //     .unwrap()
        //     .into();
        //
        // assert_snapshot!(format!("{:?}", config));
    }

    #[test]
    fn test_config() {
        use crossterm::event::KeyModifiers;

        use key_binding::{Command, Key};

        let dummy_toml = r#"
        [global]
        key_bindings = [
         { key = "q", command = "quit" },
         { key = "ctrl+g", command = "vault_selector_modal_toggle" },
         { key = "?", command = "help_modal_toggle" },
        ]
    "#;
        let dummy_toml_config: TomlConfig = toml::from_str::<TomlConfig>(dummy_toml).unwrap();

        let expected_toml_config = TomlConfig {
            global: TomlConfigSection {
                key_bindings: [
                    (Key::from('q'), Command::Quit),
                    (
                        Key::from(('g', KeyModifiers::CONTROL)),
                        Command::VaultSelectorModalToggle,
                    ),
                    (Key::from('?'), Command::HelpModalToggle),
                ]
                .into(),
            },
            ..Default::default()
        };

        assert_eq!(dummy_toml_config, expected_toml_config);

        let expected_config = Config::default().merge(expected_toml_config.into());

        assert_eq!(
            Config::default().merge(Config::from(dummy_toml_config)),
            expected_config
        );
    }
}
