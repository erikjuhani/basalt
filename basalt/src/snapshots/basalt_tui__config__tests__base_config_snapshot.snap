---
source: basalt/src/config.rs
expression: "format!(\"{:?}\", config)"
---
Config { global: ConfigSection { key_bindings: {"?": HelpModal(Toggle), "control-g": VaultSelectorModal(Toggle), "q": Quit} }, splash: ConfigSection { key_bindings: {"down": Splash(Down), "j": Splash(Down), "k": Splash(Up), "return": Splash(Open), "up": Splash(Up)} }, explorer: ConfigSection { key_bindings: {"control-b": Explorer(Toggle), "control-d": Explorer(ScrollDown(HalfPage)), "control-u": Explorer(ScrollUp(HalfPage)), "down": Explorer(Down), "j": Explorer(Down), "k": Explorer(Up), "return": Explorer(Open), "s": Explorer(Sort), "t": Explorer(Toggle), "tab": Explorer(SwitchPane), "up": Explorer(Up)} }, help_modal: ConfigSection { key_bindings: {"control-d": HelpModal(ScrollDown(HalfPage)), "control-u": HelpModal(ScrollUp(HalfPage)), "down": HelpModal(ScrollDown(One)), "esc": HelpModal(Close), "j": HelpModal(ScrollDown(One)), "k": HelpModal(ScrollUp(One)), "up": HelpModal(ScrollUp(One))} }, note_viewer: ConfigSection { key_bindings: {"control-b": NoteViewer(ToggleExplorer), "control-d": NoteViewer(ScrollDown(HalfPage)), "control-u": NoteViewer(ScrollUp(HalfPage)), "down": NoteViewer(ScrollDown(One)), "j": NoteViewer(ScrollDown(One)), "k": NoteViewer(ScrollUp(One)), "t": NoteViewer(ToggleExplorer), "tab": NoteViewer(SwitchPane), "up": NoteViewer(ScrollUp(One))} }, vault_selector_modal: ConfigSection { key_bindings: {"down": VaultSelectorModal(Down), "esc": VaultSelectorModal(Close), "j": VaultSelectorModal(Down), "k": VaultSelectorModal(Up), "return": VaultSelectorModal(Select), "up": VaultSelectorModal(Up)} } }
